### PR TITLE
Refactor dist, azl macros

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./SPECS/LICENSES-AND-NOTICES/data/licenses.json`, `./SPECS/LICENSES-AND-NOTICES/LICENSES-MAP.md`, `./SPECS/LICENSES-AND-NOTICES/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

We define `%dist` as `.azl3` (previously `.cm2`) in `/usr/lib/rpm/macros.d/dist.macros`. This is currently part of `mariner-rpm-macros` package. This has a few issues:
- mariner-rpm-macros is not included in many images, it is generally only pulled in with `rpm-build` etc.
- `%dist` can't be used easily for conditional flags in spec files (`%if 0%{?rhel}` is a common pattern).

Define a new macro `%azl 3` (which will be defined in `macros.dist` in the future). Since the `DIST_TAG` can be customized we set this macro explicitly from values baked into the go tools.

https://microsoft.sharepoint.com/:w:/t/Mariner/EQ5GV3mMEZRLsLNIwxFXj-gBxBalfkJpba-xHiz-iOJc-Q?e=Dw2Min

###### Change Log  

- Add `%azl` to build macros
- Update specs that are using similar macros to use the new versions.

###### Does this affect the toolchain?  

**YES**

###### Test Methodology

- Pipeline build id: ~https://dev.azure.com/mariner-org/mariner/_build/results?buildId=510616&view=results~ https://dev.azure.com/mariner-org/mariner/_build/results?buildId=515509&view=results

### CODE REVIEW

File: SPECS/mariner-rpm-macros/mariner-rpm-macros.spec
Comment (PawelWMS) [inline SPECS/mariner-rpm-macros/mariner-rpm-macros.spec:91]: Just an idea to keep things shorter, not an actual issue:
```suggestion
sed -e 's|@DIST@|%{dist}|g' -e "s|@MAJORVER@|$major_version|g" %{SOURCE28} > macros.dist
```

File: SPECS/systemd/systemd.spec
Comment (PawelWMS) [inline SPECS/systemd/systemd.spec:3]: Thanks for the clean-up!

File: toolkit/scripts/rpmops.sh
Comment (PawelWMS) [inline toolkit/scripts/rpmops.sh:17]: I think we may need something similar inside a few more places where we manually define the macros:
- `toolkit/scripts/toolchain/check_manifests.sh`,
- `toolkit/scripts/toolchain/build_official_toolchain_rpms.sh`.

File: toolkit/scripts/rpmops.sh
Comment (PawelWMS) [inline toolkit/scripts/rpmops.sh:31]: Please don't echo anything extra to `stdout` for a successful run - this is meant to mimic `rpmspec` and is used by tooling parsing its output.

Comment (dmcilvaney) [reply toolkit/scripts/rpmops.sh]: Sorry, I had cleaned that up locally but my git upstream got messed up, re-pushed.

**Comment by ddstreetmicrosoft** [comment]
I would really prefer my PR over this one #8028

**Comment by ddstreetmicrosoft** [comment]
> I would really prefer my PR over this one #8028

specifically, i don't want to see the mariner-rpm-macros package get *bigger*, it should get split up into separate packages. And the macros.dist file absolutely should belong to the azurelinux-release package, not mariner-rpm-macros (or a subpackage).

File: toolkit/scripts/rpmops.sh
Comment (dmcilvaney) [reply toolkit/scripts/rpmops.sh:17]: Hmm, I realize now that the toolchain call to srpmpacker needs these as well

Comment (dmcilvaney) [reply toolkit/scripts/rpmops.sh:17]: Ok, we need to define these explicitly in the go tools. I noticed we only call `DefaultDefines()` in one place (srpmpacker), and we use the distro tag version everywhere else. I unified the two into a single function and removed the special case handling from srpmpacker.

**Comment by dmcilvaney** [comment]
> > I would really prefer my PR over this one #8028
> 
> specifically, i don't want to see the mariner-rpm-macros package get _bigger_, it should get split up into separate packages. And the macros.dist file absolutely should belong to the azurelinux-release package, not mariner-rpm-macros (or a subpackage).

I'm happy to merge these changes together somehow, I'm not married to the exact implementation. Especially given I've now realized that the tools will need to define these macros independently of the actual macro file (at least for the toolchain build) the exact location of the macros isn't important to me anymore.

**Comment by dmcilvaney** [comment]
@ddstreetmicrosoft Ok, what I'll do is remove my changes to `macros.dist`. The build tools will be setting the values of these macros anyways, so the exact location/source of the file isn't critical. 

The one wrinkle there is that we need to set all the macros we care about inside the tools. How many of the new dist macros are critical? I'll leave `azl3 1` and `azl 3` defined (unless you are really keen on `azure`?), but we might need to add more to the go tools in the future.

**Comment by PawelWMS** [comment]
> > > I would really prefer my PR over this one #8028
> > 
> > 
> > specifically, i don't want to see the mariner-rpm-macros package get _bigger_, it should get split up into separate packages. And the macros.dist file absolutely should belong to the azurelinux-release package, not mariner-rpm-macros (or a subpackage).
> 
> I'm happy to merge these changes together somehow, I'm not married to the exact implementation. Especially given I've now realized that the tools will need to define these macros independently of the actual macro file (at least for the toolchain build) the exact location of the macros isn't important to me anymore.

I think I also like the idea of `macros.dist` coming from `azurelinux-release`. If we're going to make that switch, however, we have to keep a few things in mind:
- Current "source of truth" for the `dist` macro is the `DIST_TAG` variable defined in `toolkit/scripts/build_tag.mk`. It's used during the toolchain build (when no packages are available, yet) and it determines the value of `dist` everywhere else. One of the features, that it's meant to support, is to allow non-core Azure Linux teams to build packages with their own `dist` tag. We should make sure we don't introduce regressions here.
- If we want to make use of the new macros defined in `azurelinux-release`, we'd need to either add that package to the toolchain or extend the logic inside `rpm.go:DefaultDefines()` to make sure they are applied during packing of the SRPMs. Ideally, I'd prefer to rely on the `macros.dist` and not toolkit's magic as we do now, but that's a lot bigger tooling change, which I doubt we'll be able to handle right now. Or at least I wouldn't try to bundle it all up into one change.

With that in mind and Daniel's revert of the `mariner-rpm-macros.spec` changes in this PR my suggestion is:
- Take this PR first, since it's a smaller change, which will at least help us establish one version of the `azl`/`azure` macros across our specs.
- Rebase the `azurelinux-release` PR on top of this change, since it's more involved considering the two points I mentioned above.

What do you think?

**Comment by ddstreetmicrosoft** [comment]
> > > > I would really prefer my PR over this one #8028
> > > 
> > > 
> > > specifically, i don't want to see the mariner-rpm-macros package get _bigger_, it should get split up into separate packages. And the macros.dist file absolutely should belong to the azurelinux-release package, not mariner-rpm-macros (or a subpackage).
> > 
> > 
> > I'm happy to merge these changes together somehow, I'm not married to the exact implementation. Especially given I've now realized that the tools will need to define these macros independently of the actual macro file (at least for the toolchain build) the exact location of the macros isn't important to me anymore.
> 
> I think I also like the idea of `macros.dist` coming from `azurelinux-release`. If we're going to make that switch, however, we have to keep a few things in mind:
> 
> * Current "source of truth" for the `dist` macro is the `DIST_TAG` variable defined in `toolkit/scripts/build_tag.mk`. It's used during the toolchain build (when no packages are available, yet) and it determines the value of `dist` everywhere else. One of the features, that it's meant to support, is to allow non-core Azure Linux teams to build packages with their own `dist` tag. We should make sure we don't introduce regressions here.
> * If we want to make use of the new macros defined in `azurelinux-release`, we'd need to either add that package to the toolchain or extend the logic inside `rpm.go:DefaultDefines()` to make sure they are applied during packing of the SRPMs. Ideally, I'd prefer to rely on the `macros.dist` and not toolkit's magic as we do now, but that's a lot bigger tooling change, which I doubt we'll be able to handle right now. Or at least I wouldn't try to bundle it all up into one change.
> 
> With that in mind and Daniel's revert of the `mariner-rpm-macros.spec` changes in this PR my suggestion is:
> 
> * Take this PR first, since it's a smaller change, which will at least help us establish one version of the `azl`/`azure` macros across our specs.

If this is *only* changing the toolkit then go for it, as I said I don't care about changes to the toolkit. Please remove the systemd.spec changes though.

> * Rebase the `azurelinux-release` PR on top of this change, since it's more involved considering the two points I mentioned above.
> 
> What do you think?

**Comment by ddstreetmicrosoft** [comment]
> > @ddstreetmicrosoft Ok, what I'll do is remove my changes to `macros.dist`. The build tools will be setting the values of these macros anyways, so the exact location/source of the file isn't critical.
> 
> I don't particularly care what the toolkit does, as long as it results in rpms that are identical to what a 'normal' (i.e. non-toolkit) build produces.
> 
> > The one wrinkle there is that we need to set all the macros we care about inside the tools. How many of the new dist macros are critical? I'll leave `azl3 1` and `azl 3` defined (unless you are really keen on `azure`?), but we might need to add more to the go tools in the future.
> 
> %azure only please. No %azl nor %azl3.

To clarify - %azl seems like we're doing the acronym thing like %rhel, but while %rhel shortens 'red hat enterprise linux' and there isn't really any part of that that makes sense standalone (e.g. %redhat or %enterprise wouldn't really make sense as far as referring to a specific product), we have 'azure linux' so %azl isn't even acronymming correctly, and we also can simply use %azure like %fedora does - e.g. 'fedora linux' uses %fedora' and 'azure linux' can use %azure. That's what makes the most sense to me. But if we really don't want to use %azure then just %azl is the next best option IMHO. For sure it makes no sense at all to have %azl3.

Guess we should also see what @christopherco @jslobodzian think

**Comment by dmcilvaney** [comment]
I've... broken GitHub...
Commits on my branch aren't showing up.

**Comment by ddstreetmicrosoft** [comment]
> I feel like `%azure` is too broad, but that's just my personal feelings. I'll get an opinion from Chris/Jon and go from there.
> 
> As for `azl3`, I based my macros on https://docs.fedoraproject.org/en-US/packaging-guidelines/DistTag/#_conditionals, specifically the fedora `%{fc#}` and `%{el#}`. Probably not super critical for our core packages since we split our releases into separate branches today, but it would be handy for down stream teams building packages, or maybe in the future if we split our packages away. This allows a single spec to easily target multiple versions.

I don't have any problem with defining %azl3, I just think *using* it is a mistake in 99% of cases (instead, the unversioned %azure or %azl macro should be used).

**Comment by dmcilvaney** [comment]
> I don't have any problem with defining %azl3, I just think _using_ it is a mistake in 99% of cases (instead, the unversioned %azure or %azl macro should be used).

100% agree, seems very specialized and personally I'm always going to recommend using the numerical value of `azl` ( ~`%if %{azl} > 2` ).

If we want to omit it I'm certainly not too hung up on it, if you don't see any real value I'll go ahead and remove it. Any maybe by the time I'm done GH will fix itself...

**Comment by dmcilvaney** [comment]
Had a chat, going with `azl`, but I'll drop `azl#`.

**Review by ddstreetmicrosoft — APPROVED** [review]
LGTM but probably should get a review from someone else more familiar with the toolkit also

File: toolkit/scripts/rpmops.sh
Comment (PawelWMS) [inline toolkit/scripts/rpmops.sh:20]:
```suggestion
# Dist tag will be of the form '.<distro><ver>', e.g. '.azl3'., relase number will be of the form '3.0'.
```

File: toolkit/scripts/toolchain/check_manifests.sh
Comment (PawelWMS) [inline toolkit/scripts/toolchain/check_manifests.sh:28]: Now we're changing the behaviour here a bit. We're starting with `--define="with_check 0"` and then pass `--define "with_check 1"` insice `distro_defines`. I'm not sure, if anything may break but if we're making that change anyway, I'm curious, if we could switch to calling `mariner_rpmspec` altogether. This way the logic setting the defines would be kept in one place less.

File: toolkit/tools/internal/rpm/rpm.go
Comment (PawelWMS) [inline toolkit/tools/internal/rpm/rpm.go:251]:
```suggestion
// the distro tags like '%dist', '%azl'.
```

File: toolkit/tools/internal/rpm/rpm.go
Comment (PawelWMS) [inline toolkit/tools/internal/rpm/rpm.go:266]: I'm not convinced about this part for these reasons:
- We extract `distroName` from `distTag`, so there's no guarantee it's always going to match the hard-coded value of `azl` in our specs.
- We expect the distro name and version to be set correctly for our specs to work, so the warning log is too weak in my opinion.

Ideas:
1. We could turn the warning into an error. That would be a lot of pluming changes in the code, however, and it would not help with making sure we always land with `azl` as the distro name. And I'd like to avoid hard-coding `azl`, since that would create another place, which would require an update in case of any re-branding changes in the future.
2. Since `distTag`, `distroName` and `distroVersion` are something we want to control globally for the whole repository (right now from the toolkit with a potential switch to reading the values provided by a package providing `system-release`), this looks similar to what we're doing with `exe.go::ToolkitVersion`. What do you think about compiling `distTag`, `distroName` and `distroVersion` into our tooling and completely dropping the `distTag` argument from `DefaultDistroDefines`? This would require either splitting `DIST_TAG` inside our makefiles a bit or parsing it to extract the other values like you do it here (but also immediately failing in case the format is incorrect).

Comment (dmcilvaney) [reply toolkit/tools/internal/rpm/rpm.go:266]: I think we still want to have DIST_TAG configurable, but the other macros I agree can be hard-coded.

File: toolkit/scripts/build_tag.mk
Comment (PawelWMS) [inline toolkit/scripts/build_tag.mk:43]: Consistency nit-pick: you use "ABRV", "MACRO" interchangeably making it a bit confusing to understand what's going on. Also, I think the comment above `DIST_TAG` is not adding any more information, since it's only visible here right above the actual value for `DIST_TAG`. Saying `<ABRV><MAJOR_ID>` in the comments while passing `$(DIST_NAME_MACRO)$(DIST_MAJOR_ID)` in reality adds to the confusion in my opinion.

What do you think about this:
```suggestion
##help:var:DIST_NAME_ABRV:<dist_name_abrv>=Basis of the distro macros. The abbreviation of the distro name and the major version are encoded into the go tools so that they can be used to create the distro specific macro '<DIST_NAME_ABRV> <dist_major_version_number>', where 'dist_major_version_number' is the major version number extracted from '<RELEASE_MAJOR_ID>'.
DIST_NAME_ABRV            ?= azl
dist_major_version_number := $(shell echo $(RELEASE_MAJOR_ID) | cut -d'.' -f1)
dist_version_macro        := $(DIST_NAME_ABRV) $(dist_major_version_number)

##help:var:DIST_TAG:<dist_tag>=Distribution tag, defines the "dist" macro used by the specs. Default: con.
DIST_TAG            ?= .$(DIST_NAME_ABRV)$(DIST_MAJOR_ID)
```

**NOTE**: I've re-named `distro_macro` to `dist_version_macro` to avoid confusion with `DIST_TAG` and use "dist" consistently.

File: toolkit/scripts/toolchain/check_manifests.sh
Comment (PawelWMS) [inline toolkit/scripts/toolchain/check_manifests.sh:11]: Are we still using `ARCH`? It looks like we're still passing it as an argument.
```suggestion
DISTRO_MACRO=$5
ARCH=$6
```

File: toolkit/scripts/toolchain/check_manifests.sh
Comment (PawelWMS) [inline toolkit/scripts/toolchain/check_manifests.sh:17]: `DIST_TAG` only has the value, not the name of the macro.
```suggestion
    exclusiveArch=$(rpmspec -q $1 --define="with_check 0" --define="_sourcedir $spec_dir" --define="dist $DIST_TAG" --define="$DISTRO_MACRO" --qf="[%{EXCLUSIVEARCH} ]" --srpm 2>/dev/null)
```

File: toolkit/scripts/toolchain/check_manifests.sh
Comment (PawelWMS) [inline toolkit/scripts/toolchain/check_manifests.sh:23]:
```suggestion
    version=$(rpmspec -q $1 --define="with_check 0" --define="_sourcedir $spec_dir" --define="dist $DIST_TAG" --define="$DISTRO_MACRO" --qf="%{VERSION}" --srpm 2>/dev/null)
    rpmWithoutExtension=$(rpmspec -q $1 --define="with_check 0" --define="_sourcedir $spec_dir" --define="dist $DIST_TAG" --define="$DISTRO_MACRO" --target=$ARCH --qf="%{nvra}\n" 2>/dev/null)
```

File: toolkit/tools/internal/rpm/rpm.go
Comment (PawelWMS) [inline toolkit/tools/internal/rpm/rpm.go:98]: We could now also move the error checks inside `DefaultDistroDefines` to here, so that they run when the application gets started. This way we'd fail faster for invalid values and avoid having to push the `error` through the code.

Something along these lines:
```suggestion
	distNameAbreviation = exe.DistroNameAbreviation
	distMajorVersion    = ldFlagToPositiveInt(exe.DistroMajorVersion, "DistroMajorVersion")
	
	func ldFlagToPositiveInt(name, value string) int {
	result, err := strconv.Atoi(name)
	if err != nil || result <= 0 {
		err = fmt.Errorf("failed to parse Go ldflag '%s'='%s'. The value must be a positive integer", name, value)
		panic(err)
	}

	return result
}
```

File: toolkit/scripts/build_tag.mk
Comment (dmcilvaney) [reply toolkit/scripts/build_tag.mk:43]: I was changing a few bits here, merged your suggestions in.

File: toolkit/scripts/toolchain/check_manifests.sh
Comment (dmcilvaney) [reply toolkit/scripts/toolchain/check_manifests.sh:11]: Good catch, mangled that one during a revert I think.

File: toolkit/scripts/tools.mk
Comment (dmcilvaney) [reply toolkit/scripts/tools.mk:96]: Handled locally across all instansces

File: toolkit/tools/pkg/specreaderutils/specreaderutil.go
Comment (dmcilvaney) [inline toolkit/tools/pkg/specreaderutils/specreaderutil.go:299]:
```suggestion
```

File: toolkit/scripts/build_tag.mk
Comment (dmcilvaney) [inline toolkit/scripts/build_tag.mk:41]:
```suggestion
##help:var:DIST_TAG:<dist_tag>=Distribution tag, defines the "dist" macro used by the specs. Default: '.<DIST_NAME_ABRV><dist_major_version_number>' e.g., ".azl3"
```

**Review by cwize1 — APPROVED** [review]

File: toolkit/tools/internal/rpm/rpm.go
Comment (PawelWMS) [inline toolkit/tools/internal/rpm/rpm.go:106]: We're not setting or updating these values here, so no need to return them. I think we might as well completely drop this function and just move the `if majorVersion < 0 || nameAbreviation == ""` directly to the caller, since there's only one. Alternatively, the function could return a boolean and leave it to the caller to construct a proper error message.

Comment (dmcilvaney) [reply toolkit/tools/internal/rpm/rpm.go:106]: Both `loadLdDistroFlags()` and `SetDistroMacros()` call it, but I agree it doesn't need to return the values (hold over from when it was calling atoi).

File: toolkit/tools/internal/rpm/rpm_test.go
Comment (PawelWMS) [inline toolkit/tools/internal/rpm/rpm_test.go:205]: Would it be possible to test one scenario per test function? It's going to be clearer what we're checking for and what exactly failed in case of a failure.

Comment (PawelWMS) [reply toolkit/tools/internal/rpm/rpm_test.go:205]: Basically, instead of a comment line for each test case in this one function, my suggestion is to have a separate test function with a name that reflects what it's testing.

**Review by PawelWMS — APPROVED** [review]

**Pull request merged**
